packages/flutter/lib/src/widgets/two_dimensional_viewport.dart

```diff
@@ -0,0 +1,1328 @@
+  /// it if it already exists.
+  RenderBox? buildOrObtainChildFor(ChildVicinity vicinity) {
+    assert(vicinity != ChildVicinity.invalid);
+    if (_leadingXIndex == null || _trailingXIndex == null || _leadingXIndex == null || _trailingYIndex == null) {
+      // First child of this layout pass. Set leading and trailing trackers.
+      _leadingXIndex = vicinity.xIndex;
+      _trailingXIndex = vicinity.xIndex;
+      _leadingYIndex = vicinity.yIndex;
+      _trailingYIndex = vicinity.yIndex;
+    } else {
+      // If any of these are still null, we missed a child.
+      assert(_leadingXIndex != null);
+      assert(_trailingXIndex != null);
+      assert(_leadingYIndex != null);
+      assert(_trailingYIndex != null);
+
+      // Update as we go.
+      _leadingXIndex = math.min(vicinity.xIndex, _leadingXIndex!);
+      _trailingXIndex = math.max(vicinity.xIndex, _trailingXIndex!);
+      _leadingYIndex = math.min(vicinity.yIndex, _leadingYIndex!);
+      _trailingYIndex = math.max(vicinity.yIndex, _trailingYIndex!);
+    }
+    if (_needsChildRebuild || !_children.containsKey(vicinity)) {
+      invokeLayoutCallback<BoxConstraints>((BoxConstraints _) {
+        _childManager._buildChild(vicinity);
+      });
+    } else {
+      _childManager._reuseChild(vicinity);
+    }
+    if (!_children.containsKey(vicinity)) {
+      // There is no child for this vicinity, we may have reached the end of the
+      // children in one or both of the x/y indices.
+      return null;
+    }
+
+    assert(_children.containsKey(vicinity));
+    final RenderBox child = _children[vicinity]!;
+    parentDataOf(child).vicinity = vicinity;
+    return child;
+  }
+
+  /// Called after [layoutChildSequence] to compute the
+  /// [TwoDimensionalViewportParentData.paintOffset] and
+  /// [TwoDimensionalViewportParentData._paintExtent] of the child.
+  void updateChildPaintData(RenderBox child) {
+    final TwoDimensionalViewportParentData childParentData = parentDataOf(child);
+    assert(
+      childParentData.layoutOffset != null,
+      'The child with ChildVicinity(xIndex: ${childParentData.vicinity.xIndex}, '
+      'yIndex: ${childParentData.vicinity.yIndex}) was not provided a '
+      'layoutOffset. This should be set during layoutChildSequence, '
+      'representing the position of the child.'
+    );
+    assert(child.hasSize); // Child must have been laid out by now.
+
+    // Set paintExtent (and visibility)
+    childParentData._paintExtent = computeChildPaintExtent(
+      childParentData.layoutOffset!,
+      child.size,
+    );
+    // Set paintOffset
+    childParentData.paintOffset = computeAbsolutePaintOffsetFor(
+      child,
+      layoutOffset: childParentData.layoutOffset!,
+      paintExtent: childParentData._paintExtent,
+    );
+    // If the child is partially visible, or not visible at all, there is
+    // visual overflow.
+    _hasVisualOverflow = childParentData.layoutOffset != childParentData._paintExtent
```

Comment:
nit: for readability I would move `_hasVisualOverflow` all the way to the front of this `or` chain.